src/vm/database/sqlite.rs

```diff
@@ -0,0 +1,305 @@
+                      (contract_identifier INTEGER PRIMARY KEY AUTOINCREMENT,
+                       contract_name TEXT,
+                       contract_data TEXT)",
+                            NO_PARAMS)?;
+
+        contract_db.check_schema()?;
+
+        Ok(contract_db)
+    }
+
+    pub fn open(filename: &str) -> Result<SqliteContractDatabase> {
+        let contract_db = SqliteContractDatabase::inner_open(filename)?;
+
+        contract_db.check_schema()?;
+        Ok(contract_db)
+    }
+
+    fn check_schema(&self) -> Result<()> {
+        let sql = "SELECT sql FROM sqlite_master WHERE name=?";
+        if let Some(ref conn) = self.conn {
+            let _: String = conn.query_row(sql, &["maps_table"],
+                                           |row| row.get(0))
+                .map_err(|x| Error::new(ErrType::SqliteError(IncomparableError{ err: x })))?;
+            let _: String = conn.query_row(sql, &["contracts"],
+                           |row| row.get(0))
+                .map_err(|x| Error::new(ErrType::SqliteError(IncomparableError{ err: x })))?;
+            let _: String = conn.query_row(sql, &["data_table"],
+                           |row| row.get(0))
+                .map_err(|x| Error::new(ErrType::SqliteError(IncomparableError{ err: x })))?;
+            Ok(())
+        } else {
+            Err(Error::new(ErrType::SqlConnectionClosed))
+        }
+    }
+
+    fn inner_open(filename: &str) -> Result<SqliteContractDatabase> {
+        let conn = Connection::open(filename)
+            .map_err(|x| Error::new(ErrType::SqliteError(IncomparableError{ err: x })))?;
+        Ok(SqliteContractDatabase {
+            conn: Some(conn),
+            save_point: 0
+        })
+    }
+
+    fn load_map(&self, contract_name: &str, map_name: &str) -> Result<SqliteDataMap> {
+        let (map_identifier, key_type, value_type): (_, String, String) =
+            self.query_row(
+                "SELECT map_identifier, key_type, value_type FROM maps_table WHERE contract_name = ? AND map_name = ?",
+                &[contract_name, map_name],
+                |row| {
+                    (row.get(0), row.get(1), row.get(2))
+                })?
+            .ok_or(Error::new(ErrType::UndefinedMap(map_name.to_string())))?;
+
+        Ok(SqliteDataMap {
+            map_identifier: map_identifier,
+            key_type: TypeSignature::deserialize(&key_type)?,
+            value_type: TypeSignature::deserialize(&value_type)?
+        })
+    }
+
+    fn load_contract(&self, contract_name: &str) -> Result<Option<Contract>> {
+        let contract: Option<String> =
+            self.query_row(
+                "SELECT contract_data FROM contracts WHERE contract_name = ?",
+                &[contract_name],
+                |row| {
+                    row.get(0)
+                })?;
+        match contract {
+            None => Ok(None),
+            Some(ref contract) => Ok(Some(Contract::deserialize(contract)?))
```

Comment:
Should this `panic!` if the database has a corrupt contract in it?

Reply:
Yep -- changed this to panic.